File: Console/Classes/ASL/ASL.m
/*******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015 Jean-David Gadina - www-xs-labs.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

#import "ASL.h"
#import "ASLMessage.h"
#import "ASLSender.h"

@import asl;

@interface ASL()

@property( atomic, readwrite, strong ) NSArray< ASLMessage * > * messages;
@property( atomic, readwrite, strong ) NSArray< ASLSender  * > * senders;

@property( atomic, readwrite, strong ) NSString * sender;
@property( atomic, readwrite, assign ) BOOL       run;
@property( atomic, readwrite, assign ) BOOL       runing;
@property( atomic, readwrite, assign ) BOOL       exit;
@property( atomic, readwrite, assign ) BOOL       inited;

- ( void )processMessages;
- ( ASLSender * )senderWithName: ( NSString * )name facility: ( NSString * )facility;

@end

@implementation ASL

- ( instancetype )init
{
    return [ self initWithSender: nil ];
}

- ( instancetype )initWithSender: ( nullable NSString * )sender
{
    if( ( self = [ super init ] ) )
    {
        self.sender   = sender;
        self.messages = @[];
        self.senders  = @[];
        
        [ NSThread detachNewThreadSelector: @selector( processMessages ) toTarget: self withObject: nil ];
    }
    
    return self;
}

- ( void )dealloc
{
    self.exit = YES;
    
    while( self.runing );
}

- ( void )start
{
    self.run = YES;
}

- ( void )stop
{
    self.run = NO;
}

- ( void )processMessages
{
    NSUInteger       lastID;
    aslclient        client;
    
    @autoreleasepool
    {
        lastID = self.messages.lastObject.messageID;
        client = asl_open( NULL, NULL, 0 );
        
        while( self.exit == NO )
        {
            self.runing = YES;
            
            {
                aslmsg           query;
                aslresponse      response;
                aslmsg           msg;
                ASLMessage     * message;
                ASLSender      * sender;
                NSMutableArray * messages;
                NSMutableArray * senders;
                
                query = asl_new( ASL_TYPE_QUERY );
                
                if( self.sender )
                {
                    asl_set_query( query, ASL_KEY_MSG, self.sender.UTF8String, ASL_QUERY_OP_EQUAL );
                }
                
                asl_set_query( query, ASL_KEY_MSG, NULL, ASL_QUERY_OP_NOT_EQUAL );
                asl_set_query( query, ASL_KEY_MSG_ID, [ NSString stringWithFormat: @"%lu", lastID ].UTF8String, ASL_QUERY_OP_GREATER );
                
                response = asl_search( client, query );
                
                asl_free( query );
                
                msg = asl_next( response );
                
                if( msg != NULL )
                {
                    messages = [ self.messages mutableCopy ];
                    
                    while( msg )
                    {
                        message = [ [ ASLMessage alloc ] initWithASLMessage: msg ];
                        lastID  = message.messageID;
                        
                        [ messages addObject: message ];
                        
                        msg = asl_next( response );
                    }
                    
                    senders = [ self.senders mutableCopy ];
                    
                    for( message in messages )
                    {
                        sender = nil;
                        
                        for( sender in senders )
                        {
                            if( [ sender.name isEqualToString: message.sender ] )
                            {
                                break;
                            }
                        }
                        
                        if( sender == nil )
                        {
                            sender = [ [ ASLSender alloc ] initWithName: message.sender facility: message.facility ];
                            
                            [ senders addObject: sender ];
                        }
                        
                        [ sender addMessage: message ];
                    }
                    
                    dispatch_sync
                    (
                        dispatch_get_main_queue(),
                        ^( void )
                        {
                            self.messages = messages;
                            self.senders  = senders;
                        }
                    );
                }
                
                asl_release( response );
                
                [ NSThread sleepForTimeInterval: 1 ];
            }
        }
        
        asl_close( client );
        
        self.runing = NO;
    }
}

- ( ASLSender * )senderWithName: ( NSString * )name facility: ( NSString * )facility
{
    ASLSender      * sender;
    NSMutableArray * senders;
    
    for( sender in self.senders )
    {
        if( [ sender.name isEqualToString: name ] && [ sender.facility isEqualToString: facility ] )
        {
            return sender;
        }
    }
    
    sender  = [ [ ASLSender alloc ] initWithName: name facility: facility ];
    senders = [ self.senders mutableCopy ];
    
    [ senders addObject: sender ];
    
    dispatch_sync
    (
        dispatch_get_main_queue(),
        ^( void )
        {
            self.senders = senders;
        }
    );
    
    return sender;
}

@end

File: Console/Classes/ASL/ASL.swift
/*******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2017 Jean-David Gadina - www.xs-labs.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

import Cocoa
import asl

@objc class ASL: NSObject
{
    @objc public private( set ) dynamic var messages: [ ASLMessage ] = []
    @objc public private( set ) dynamic var senders:  [ ASLSender ]  = []
    
    @objc private dynamic var sender: String = ""
    @objc private dynamic var run:    Bool   = false
    @objc private dynamic var runing: Bool   = false
    @objc private dynamic var exit:   Bool   = false
    @objc private dynamic var inited: Bool   = false
    
    @objc convenience override init()
    {
        self.init( sender: nil )
    }
    
    @objc init( sender: String? )
    {
        super.init()
        
        self.sender = sender ?? ""
        
        Thread.detachNewThreadSelector( #selector( processMessages ), toTarget: self, with: nil )
    }
    
    deinit
    {
        for sender in self.senders
        {
            sender.removeObserver( self, forKeyPath: "messages" )
        }
        
        self.exit = true
        
        while self.runing
        {}
    }
    
    override func observeValue( forKeyPath keyPath: String?, of object: Any?, change: [ NSKeyValueChangeKey : Any ]?, context: UnsafeMutableRawPointer? )
    {
        if( keyPath == "messages" && object is ASLSender )
        {
            DispatchQueue.main.async
            {
                if( self.senders.contains( object as! ASLSender ) )
                {
                    var allMessages = [ ASLMessage ]()
                    
                    for sender in self.senders
                    {
                        allMessages.append( contentsOf: sender.messages as! [ ASLMessage ] )
                    }
                    
                    self.messages = allMessages
                }
            }
        }
        else
        {
            super.observeValue( forKeyPath: keyPath, of: object, change: change, context: context )
        }
    }
    
    @objc public func start()
    {
        self.run = true
    }
    
    @objc public func stop()
    {
        self.run = false
    }
    
    @objc private func processMessages()
    {
        var lastID = self.messages.last?.messageID ?? 0
        let client = asl_open( nil, nil, 0 )
        
        while( self.exit == false )
        {
            self.runing = true
            
            let query = asl_new( UInt32( ASL_TYPE_QUERY ) )
            
            if( self.sender.count > 0 )
            {
                asl_set_query( query, ASL_KEY_MSG, ( self.sender as NSString ).utf8String, UInt32( ASL_QUERY_OP_EQUAL ) )
            }
            
            asl_set_query( query, ASL_KEY_MSG, nil, UInt32( ASL_QUERY_OP_NOT_EQUAL ) )
            asl_set_query( query, ASL_KEY_MSG_ID, ( String( format: "%lu", lastID ) as NSString ).utf8String, UInt32( ASL_QUERY_OP_GREATER ) )
            
            let response = asl_search( client, query )
            
            asl_free( query )
            
            var msg = asl_next( response )
            
            if( msg != nil )
            {
                var newMessages = [ ASLMessage ]()
                
                while( msg != nil )
                {
                    let message = ASLMessage( msg )
                    
                    if( message.messageID <= lastID )
                    {
                        continue
                    }
                    
                    lastID = ( message.messageID > lastID ) ? message.messageID : lastID
                    
                    newMessages.append( message )
                    
                    msg = asl_next( response )
                }
                
                var senders = [ ASLSender ]( self.senders )
                
                for message in newMessages
                {
                    var sender: ASLSender?
                    
                    for s in senders
                    {
                        if( s.name == message.sender && s.facility == message.facility )
                        {
                            sender = s
                            
                            break
                        }
                    }
                    
                    if( sender == nil )
                    {
                        sender = ASLSender( name: message.sender, facility: message.facility )
                        
                        senders.append( sender! )
                        sender?.addObserver( self, forKeyPath: "messages", options: .new, context: nil )
                    }
                    
                    sender?.addMessage( message )
                }
                
                var allMessages: [ ASLMessage ]?
                
                if( newMessages.count == 0 )
                {
                    allMessages = nil
                }
                else
                {
                    allMessages = []
                    
                    for sender in senders
                    {
                        allMessages?.append( contentsOf: sender.messages as! [ ASLMessage ] )
                    }
                }
                
                DispatchQueue.main.sync
                {
                    if( allMessages != nil )
                    {
                        self.messages = allMessages!
                    }
                    
                    self.senders = senders
                }
            }
        }
    }
    
    private func sender( with name: String, facility: String ) -> ASLSender
    {
        for sender in self.senders
        {
            if( sender.name == name && sender.facility == facility )
            {
                return sender
            }
        }
        
        let sender  = ASLSender( name: name, facility: facility )
        var senders = [ ASLSender ]( self.senders )
        
        senders.append( sender )
        
        DispatchQueue.main.sync
        {
            self.senders = senders
        }
        
        return sender
    }
}
